src/web100srv.c

```diff
@@ -1366,90 +1266,64 @@ void process_parent_message(int parent_message, Connection *ctl,
   // runs the actual tests.
-  log_println(4, "Child thinks pipe() returned fd0=%d, fd1=%d for pid=%d",
-              chld_pipe[0], chld_pipe[1], chld_pid);
-  close(listenfd);
-  close(chld_pipe[1]);
+  // First, start the watchdog timer
+  alarm(300);
+
+  // Set up the connection with the client (with optional SSL and websockets)
+  if (options.tls) {
+    if (setup_SSL_connection(&ctl, ssl_context) != 0) {
+      log_println(1, "setup_SSL_connection failed");
+      exit(-1);
+    }
+  }
+
+  // Read the login message and send the kickoff message (if applicable)
+  t_opts = initialize_tests(&ctl, &testopt, test_suite, sizeof(test_suite));
+  if (t_opts < 1) {  // some error in initialization routines
+    log_println(3, "Invalid test suite received, terminate child");
+    exit(-1);
+  }
+
 #if USE_WEB100
   if ((agent = web100_attach(WEB100_AGENT_TYPE_LOCAL, NULL)) == NULL) {
     web100_perror("web100_attach");
-    return 1;
+    exit(1);
   }
 #elif USE_WEB10G
   if (estats_nl_client_init(&agent) != NULL) {
-    log_println(0,
-                  "Error: estats_client_init failed."
-                  "Unable to use web10g.");
-    return 1;
+    log_println(0, "Error: estats_client_init failed. Unable to use web10g.");
+    exit(1);
   }
 #endif
-  // This is the child process from the above fork().  The parent
-  //  is in control, and will send this child a signal when it gets
-  //  to the head of the testing queue.  Post a read() and simply
-  //  wait for the parent to let us know it's time to move on.
-  //  Rich Carlson 3/11/04
-  for (;;) {
-    memset(buff, 0, sizeof(buff));
-
-    // the read() could return if an interrupt was caught.  This
-    // condition should be checked for and the read() restarted if
-    // necessary
-    // RAC 3/18/10
-    //
-    retcode = read(chld_pipe[0], buff, 32);
-    log_println(6, "Child %d received '%s' from parent", getpid(),
-                buff);
-    if ((retcode == -1) && (errno == EINTR))
-      continue;
-    if (strncmp(buff, "go", 2) == 0) {
-      log_println(6, "Got 'go' signal from parent, ready to start "
-                  "testing %d", getpid());
-      break;
-    }
-    if (buff[0] == 0) {
-      log_println(6, "Child %d received 'go' signal, exiting",
-                  getpid());
-      exit(-1);
-    }
+  // Wait in the queue until the child process is told to start
+  while ((parent_message = read_from_parent(parent_pipe)) !=
+         SRV_QUEUE_TEST_STARTS_NOW) {
+    process_parent_message(parent_message, &ctl, &testopt, t_opts);
   }
 
-  set_timestamp();
+  // Tell the client the test is about to start
+  if (send_json_message_any(&ctl, SRV_QUEUE, SRV_QUEUE_TEST_STARTS_NOW_STR,
+                            strlen(SRV_QUEUE_TEST_STARTS_NOW_STR),
+                            testopt.connection_flags, JSON_SINGLE_VALUE) != 0) {
+    log_println(1, "Error when sending SRV_QUEUE_TEST_STARTS_NOW to client");
+    exit(-1);
+  }
```

Comment:
Can we refactor these queueing messages into a helper function so that this:

```
    char serialized_parent_message[256];
    sprintf(serialized_parent_message, "%d", parent_message);
    if (send_json_message_any(ctl, SRV_QUEUE, serialized_parent_message,
                              strlen(serialized_parent_message),
                              testopt->connection_flags,
                              JSON_SINGLE_VALUE) != 0) {
      log_println(1, "Client failed to send message");
      exit(0);
    }
```

Can become this:

```
    send_queue_message(ctl, parent_message, testopt->connection_flags);
```

And then we can get rid of those `SRV_QUEUE_*_STR` constants that I've had it in for the whole review.